aptos-move/framework/aptos-framework/sources/HashMap.move

```diff
@@ -4,76 +4,134 @@ module AptosFramework::HashMap {
+    /// Create an empty HashMap with `initial_buckets` buckets.
     public fun new<K: drop + store, V: store>(initial_buckets: u64): HashMap<K, V> {
-        // assert it's power of 2
-        assert!(initial_buckets > 0 && (initial_buckets & (initial_buckets - 1) == 0), 0);
+        assert!(initial_buckets > 0 && (initial_buckets & (initial_buckets - 1) == 0), Errors::invalid_argument(ENOT_POWER_OF_TWO));
         let buckets = Table::new();
         let i = 0;
+        let level = 0;
         while (i < initial_buckets) {
             Table::add(&mut buckets, &i, Vector::empty());
             i = i + 1;
+            if (i == 1 << (level + 1)) {
+                level = level + 1;
+            }
         };
         HashMap {
             buckets,
             num_buckets: initial_buckets,
+            level,
             len: 0,
         }
     }
 
     /// Destroy empty map.
     /// Aborts if it's not empty.
     public fun destroy_empty<K, V>(map: HashMap<K, V>) {
-        assert!(map.len == 0, 0);
+        assert!(map.len == 0, Errors::invalid_argument(ENOT_EMPTY));
         let i = 0;
         while (i < map.num_buckets) {
             Vector::destroy_empty(Table::remove(&mut map.buckets, &i));
             i = i + 1;
         };
-        let HashMap {buckets, num_buckets: _, len: _} = map;
+        let HashMap {buckets, num_buckets: _, level: _, len: _} = map;
         Table::destroy_empty(buckets);
     }
 
-    /// Insert (key, value) pair in the hash map.
+    /// Insert (key, value) pair in the hash map, it may grow one bucket if current load factor exceeds the threshold.
+    /// Note it may not split the actual overflowed bucket.
     /// Abort if `key` already exists.
     public fun insert<K, V>(map: &mut HashMap<K, V>, key: K, value: V) {
-        let hash = sip_hash(&key) % map.num_buckets;
-        let bucket = Table::borrow_mut(&mut map.buckets, &hash);
+        let hash = sip_hash(&key);
+        let index = bucket_index(map, hash);
+        let bucket = Table::borrow_mut(&mut map.buckets, &index);
         let i = 0;
         let len = Vector::length(bucket);
         while (i < len) {
             let entry = Vector::borrow(bucket, i);
             assert!(&entry.key != &key, 0);
             i = i + 1;
         };
-        Vector::push_back(bucket, Entry {key, value});
+        Vector::push_back(bucket, Entry {hash, key, value});
         map.len = map.len + 1;
+
+        if (load_factor(map) > SPLIT_THRESHOLD) {
+            split_one_bucket(map);
+        }
+    }
+
+    /// Split the next bucket into two and re-insert existing items.
+    fun split_one_bucket<K, V>(map: &mut HashMap<K, V>) {
+        let num_buckets = map.num_buckets;
+        // the next bucket to split is num_bucket without the most significant bit.
+        let to_split = num_buckets ^ (1 << map.level);
+        let old_bucket = Table::remove(&mut map.buckets, &to_split);
+        Table::add(&mut map.buckets, &to_split, Vector::empty());
```

Comment:
The extension potentially charge both the remove and add (and potentially even more on the add) in addition to the write.. so you might want to borrow_mut() and truncate and work from there..